### PR TITLE
[WIP] Hotfix for Null Pointer Dereference in BestPractices::PreCallValidateCreateInstance

### PR DESCRIPTION
**Pull request by samwalls-arm**
This is the same issue that came up in this PR: https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/1750

However, it's in a different place.

I've added tests here to make sure this won't happen in the future. However, currently the tests don't seem to be able to hook into PreCallValidateCreateInstance / PostCallRecordCreateInstance. I'm not sure why.

### CODE REVIEW

**Comment by mark-lunarg** [comment]
I thought we had closed the instance-time error-reporting holes, but I'll look at it again.

**Comment by mark-lunarg** [comment]
@samwalls-arm, doesn't the crash prove that the tests/validation are hooking PreCallValidateCreateInstance?

**Comment by samwalls-arm** [comment]
Hey @mark-lunarg, the crash comes from a bug report within Arm from someone using the Best Practices layer in the wild.

**Comment by samwalls-arm** [comment]
I submitted this from a windows build where I didn't have access to a proper debugger. Trying it out now on Linux with GDB, I can confirm that if I set a breakpoint in PreCallValidateCreateInstance, it's never reached when stepping over the tests here.

**Comment by mark-lunarg** [comment]
@samwalls-arm, this is puzzling, but I did rebase these changes to TOT before testing, so maybe that affected my results.  Anyway, running through CI now.  I'm assuming that the newly added file was accidental and unnecessary, yes?

**Comment by samwalls-arm** [comment]
@mark-lunarg Yes, that's a reference to an emacs buffer, I don't know why it got added!

The tests may not reflect the behaviour we want to falsify currently, as they don't run, however we should maybe include the fix itself at a minimum?

**Comment by mark-lunarg** [comment]
Ah, yeah, you'll need an `InitBestPracticesFramework `call in your test cases.  We really need to get all of that hooked up inside the `VkBestPracticesLayerTest `class, so that it's automatic somehow.

**Comment by samwalls-arm** [comment]
Oh, good spot. I can fix that now.

**Comment by samwalls-arm** [comment]
@mark-lunarg So, using `InitBestPracticesFramework` does trigger `PreCallValidateCreateInstance`, However, it only triggers for the `vkCreateInstance` call inside the test framework itself. I still can't get the test to trigger the layer with null `pApplicationInfo`. Will push changes in a moment.

**Comment by mark-lunarg** [comment]
This is causing a worrisome crash in internal CI which is under investigation.

**Comment by samwalls-arm** [comment]
@mark-lunarg How long do we have to fix the tests before the next SDK release? If it's not long, perhaps we should split the tests away into a separate PR, until we can figure out what's going on.

**Comment by mark-lunarg** [comment]
@samwalls-arm, the SDK branches were created about a week ago. I'll try to see if it's not too late to cherry-pick this fix, or part of it.

**Comment by mark-lunarg** [comment]
OK, @samwalls-arm, I pushed a commit to fix the null deref -- it's different from your first commit but accomplishes the same thing.  @Tony-LunarG is trying to find the root-cause of the very strange CI failures caused by the commit with the tests in it (!).  Working on getting it cherry-picked over now.

File: layers/best_practices_utils.cpp [hunk 1]
```diff
@@ -132,16 +132,17 @@ bool BestPractices::ValidateDeprecatedExtensions(const char* api_name, const cha
...
         if (white_list(pCreateInfo->ppEnabledExtensionNames[i], kDeviceExtensionNames)) {
             skip |= LogWarning(instance, kVUID_BestPractices_CreateInstance_ExtensionMismatch,
                                "vkCreateInstance(): Attempting to enable Device Extension %s at CreateInstance time.",
                                pCreateInfo->ppEnabledExtensionNames[i]);
         }
-        skip |= ValidateDeprecatedExtensions("CreateInstance", pCreateInfo->ppEnabledExtensionNames[i],
-                                             pCreateInfo->pApplicationInfo->apiVersion,
-                                             kVUID_BestPractices_CreateInstance_DeprecatedExtension);
+        if (pCreateInfo->pApplicationInfo != nullptr) {
+            skip |= ValidateDeprecatedExtensions("CreateInstance", pCreateInfo->ppEnabledExtensionNames[i],
```
Comment (mark-lunarg) [inline layers/best_practices_utils.cpp:142]: This commit is no longer necessary.

File: tests/vklayertests_best_practices.cpp [hunk 2]
```diff
@@ -92,6 +92,40 @@ TEST_F(VkBestPracticesLayerTest, ValidateReturnCodes) {
...
+    TEST_DESCRIPTION("Attempt to create a Vulkan instance with a null VkApplicationInfo. This should not cause an error.");
+
+    InitBestPracticesFramework();
+
+    VkInstanceCreateInfo ici = GetInstanceCreateInfo();
+    ici.pApplicationInfo = nullptr;
+
+    VkInstance instance;
+    m_errorMonitor->ExpectSuccess();
+    VkResult result = vk::CreateInstance(&ici, nullptr, &instance);
```
Comment (mark-lunarg) [inline tests/vklayertests_best_practices.cpp:105]: This instance gets leaked.  For Positive tests everything has to be destroyed -- instances are problematic in that if you don't destroy an instance, validation cannot tell you what didn't get cleaned up.

**Comment by mark-lunarg** [comment]
@samwalls-arm, closing this for now due to inactivity -- when there's time later, we can reopen this PR or just create a new one -- thanks!

**Pull request closed without merging**